homeassistant/components/simplisafe/__init__.py

```diff
@@ -0,0 +1,133 @@
+            vol.All(cv.ensure_list, [ACCOUNT_CONFIG_SCHEMA]),
+    }),
+}, extra=vol.ALLOW_EXTRA)
+
+
+async def async_setup(hass, config):
+    """Set up the SimpliSafe component."""
+    hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][DATA_CLIENT] = {}
+    hass.data[DOMAIN][DATA_LISTENER] = {}
+
+    if DOMAIN not in config:
+        return True
+
+    conf = config[DOMAIN]
+
+    for account in conf[CONF_ACCOUNTS]:
+        if account[CONF_USERNAME] in configured_instances(hass):
+            continue
+
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={'source': SOURCE_IMPORT},
+                data={
+                    CONF_USERNAME: account[CONF_USERNAME],
+                    CONF_PASSWORD: account[CONF_PASSWORD],
+                    CONF_CODE: account.get(CONF_CODE),
+                }))
+
+    return True
+
+
+async def async_setup_entry(hass, config_entry):
+    """Set up OpenUV as config entry."""
+    from simplipy import API
+    from simplipy.errors import SimplipyError
+
+    token_filepath = hass.config.path(
+        DATA_FILE_SCAFFOLD.format(config_entry.data[CONF_USERNAME]))
+    token_data = await hass.async_add_executor_job(load_json, token_filepath)
+
+    websession = aiohttp_client.async_get_clientsession(hass)
+
+    try:
+        simplisafe = await API.login_via_token(
+            token_data['refresh_token'], websession)
+    except SimplipyError as err:
+        _LOGGER.error("There was an error during setup: %s", err)
+        return False
+
+    await hass.async_add_executor_job(
+        save_json, token_filepath, {'refresh_token': simplisafe.refresh_token})
+
+    systems = await simplisafe.get_systems()
+    hass.data[DOMAIN][DATA_CLIENT][config_entry.entry_id] = systems
+
+    hass.async_create_task(
+        hass.config_entries.async_forward_entry_setup(
+            config_entry, 'alarm_control_panel'))
+
+    async def refresh(event_time):
+        """Refresh data from the SimpliSafe account."""
+        for system in systems:
+            _LOGGER.debug('Updating system data: %s', system.system_id)
+            await system.update()
+            async_dispatcher_send(hass, TOPIC_UPDATE.format(system.system_id))
+
+            if system.api.refresh_token_dirty:
+                await hass.async_add_executor_job(
+                    save_json, token_filepath,
+                    {'refresh_token': system.api.refresh_token})
+
+    hass.data[DOMAIN][DATA_LISTENER][
+        config_entry.entry_id] = async_track_time_interval(
+            hass, refresh,
+            config_entry.data.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL))
+
+    return True
+
+
+async def async_unload_entry(hass, entry):
+    """Unload a SimpliSafe config entry."""
+    hass.data[DOMAIN][DATA_CLIENT].pop(entry.entry_id)
+    remove_listener = hass.data[DOMAIN][DATA_LISTENER].pop(entry.entry_id)
+    remove_listener()
+
```

Comment:
Forward entry unload to all platforms here. That will unload the alarm_control_panel platform.

See deconz component for example.

Reply:
I've added that code in my latest commit; the alarm control panel still doesn't get removed.

Reply:
Is it the entity that is still left?

Reply:
Correct: the config entry (thus, the component goes away), but the alarm control panel entity remains.

Reply:
Any errors in the logs?

Reply:
Negative, unfortunately. A reboot of HASS clears away the alarm control panel, but even then, the logs don’t show anything. I’ll trace the unload calls and see if I can figure out where the break occurs.

Reply:
No break that I can see. The only strangeness (to me) are these logs (with some of my debugging statements in place):

```
1. 2018-10-10 09:58:14 DEBUG (MainThread) [homeassistant.core] Removing entity: alarm_control_panel.1234_main_street
2. 2018-10-10 09:58:14 DEBUG (MainThread) [homeassistant.core] Firing event
3. 2018-10-10 09:58:14 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=alarm_control_panel.1234_main_street, old_state=<state alarm_control_panel.1234_main_street=disarmed; code_format=None, changed_by=None, alarm_active=False, temperature=65, friendly_name=1234 Main Street @ 2018-10-10T09:57:57.711751-06:00>, new_state=None>
4. 2018-10-10 09:58:14 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection.4559956960] Sending {'id': 2, 'type': 'event', 'event': {'event_type': 'state_changed', 'data': {'entity_id': 'alarm_control_panel.1234_main_street', 'old_state': <state alarm_control_panel.1234_main_street=disarmed; code_format=None, changed_by=None, alarm_active=False, temperature=65, friendly_name=1234 Main Street @ 2018-10-10T09:57:57.711751-06:00>, 'new_state': None}, 'origin': 'LOCAL', 'time_fired': datetime.datetime(2018, 10, 10, 15, 58, 14, 595742, tzinfo=<UTC>), 'context': {'id': '1fb3f29315b44ed68b00afd99e2b3b17', 'user_id': None}}}
5. 2018-10-10 09:58:14 DEBUG (MainThread) [homeassistant.config_entries] Unload result is True
6. 2018-10-10 09:58:14 DEBUG (MainThread) [homeassistant.config_entries] Unload result is True
7. 2018-10-10 09:58:14 DEBUG (MainThread) [homeassistant.config_entries] state adjusted
8. 2018-10-10 09:58:14 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=alarm_control_panel.1234_main_street, old_state=None, new_state=<state alarm_control_panel.1234_main_street=disarmed; code_format=None, changed_by=None, alarm_active=False, temperature=65, friendly_name=1234 Main Street @ 2018-10-10T09:58:14.621691-06:00>>
9. 2018-10-10 09:58:14 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection.4559956960] Sending {'id': 2, 'type': 'event', 'event': {'event_type': 'state_changed', 'data': {'entity_id': 'alarm_control_panel.1234_main_street', 'old_state': None, 'new_state': <state alarm_control_panel.1234_main_street=disarmed; code_format=None, changed_by=None, alarm_active=False, temperature=65, friendly_name=1234 Main Street @ 2018-10-10T09:58:14.621691-06:00>}, 'origin': 'LOCAL', 'time_fired': datetime.datetime(2018, 10, 10, 15, 58, 14, 621708, tzinfo=<UTC>), 'context': {'id': '8442566714d3438fb7b00f58fc0a5739', 'user_id': None}}}
```

From what I'm reading:

* Line 1: `homeassistant.core.async_remove` is called with the correct entity ID.
* Lines 2-4: The `state_changed` event is fired (with an `old_state` of whatever the alarm is currently set at and a `new_state` of `None`).
* Lines 5-6: `component.async_unload_entry` is called twice (with the result being `True` both times).
* Line 7: `hass.config_entries.state` is set to `not_loaded`.
* Line 8: Another `state_changed` event is fired (with an `old_state` of `None` and a `new_state` of whatever the alarm is currently set at).

I'm not the expert on this subsystem, but it seems like there might be two areas of strangeness:

1. Should `component.async_unload_entry` be called twice?
2. Why does the second state change occur at the end? Does a non-`None` state somehow keep the entity around?

Apologies I'm not more famliar. Happy to help debugging if you can direct me.

Reply:
Check which component is used for generating the `unload result`.

Reply:
```
2018-10-11 09:08:58 DEBUG (MainThread) [homeassistant.config_entries] Unload called from <module 'homeassistant.components.simplisafe' from '/Users/abach/Git/home-assistant/homeassistant/components/simplisafe/__init__.py'>
2018-10-11 09:08:58 DEBUG (MainThread) [homeassistant.config_entries] Unload called from <module 'homeassistant.components.alarm_control_panel' from '/Users/abach/Git/home-assistant/homeassistant/components/alarm_control_panel/__init__.py'>
```

Reply:
FYI, tracking this issue separately: #17370.